akka-http-core/src/main/scala/akka/http/util/StreamUtils.scala

```diff
@@ -193,6 +196,125 @@ private[http] object StreamUtils {
+    new Sink[In, Publisher[In]](new OneTimePublisherSink(none, SinkShape(new Inlet(name)), cell))
+  def oneTimeSubscriberSource[Out](cell: OneTimeWriteCell[Subscriber[Out]], name: String): Source[Out, Subscriber[Out]] =
+    new Source[Out, Subscriber[Out]](new OneTimeSubscriberSource(none, SourceShape(new Outlet(name)), cell))
+
+  /** A copy of PublisherSink that allows access to the publisher through the cell but can only materialized once */
+  private class OneTimePublisherSink[In](attributes: OperationAttributes, shape: SinkShape[In], cell: OneTimeWriteCell[Publisher[In]])
+    extends PublisherSink[In](attributes, shape) {
+    override def create(materializer: ActorFlowMaterializerImpl, flowName: String): (Subscriber[In], Publisher[In]) = {
+      val results = super.create(materializer, flowName)
+      cell.set(results._2)
+      results
+    }
+    override protected def newInstance(shape: SinkShape[In]): SinkModule[In, Publisher[In]] =
+      new OneTimePublisherSink[In](attributes, shape, cell)
+
+    override def withAttributes(attr: OperationAttributes): Module =
+      new OneTimePublisherSink[In](attr, amendShape(attr), cell)
+  }
+  /** A copy of SubscriberSource that allows access to the subscriber through the cell but can only materialized once */
+  private class OneTimeSubscriberSource[Out](val attributes: OperationAttributes, shape: SourceShape[Out], cell: OneTimeWriteCell[Subscriber[Out]])
+    extends SourceModule[Out, Subscriber[Out]](shape) {
+
+    override def create(materializer: ActorFlowMaterializerImpl, flowName: String): (Publisher[Out], Subscriber[Out]) = {
+      val processor = new Processor[Out, Out] {
+        @volatile private var subscriber: Subscriber[_ >: Out] = null
+
+        override def subscribe(s: Subscriber[_ >: Out]): Unit = subscriber = s
+        override def onError(t: Throwable): Unit = subscriber.onError(t)
+        override def onSubscribe(s: Subscription): Unit = subscriber.onSubscribe(s)
+        override def onComplete(): Unit = subscriber.onComplete()
+        override def onNext(t: Out): Unit = subscriber.onNext(t)
+      }
+      cell.set(processor)
+
+      (processor, processor)
+    }
+
+    override protected def newInstance(shape: SourceShape[Out]): SourceModule[Out, Subscriber[Out]] =
+      new OneTimeSubscriberSource[Out](attributes, shape, cell)
+    override def withAttributes(attr: OperationAttributes): Module =
+      new OneTimeSubscriberSource[Out](attr, amendShape(attr), cell)
+  }
+
+  trait ReadableCell[+T] {
+    def value: T
+  }
+  /** A one time settable cell */
+  class OneTimeWriteCell[T <: AnyRef] extends ReadableCell[T] {
+    @volatile var _value: T = null.asInstanceOf[T]
+    def value: T = {
+      require(_value != null, "Value wasn't set yet")
+      _value
+    }
+
+    def set(value: T): Unit = {
+      require(_value == null, "Value can be only set once.") // one time usage only
```

Comment:
isn't a compareAndSet cleaner? Also, the error message should indicate that null values are not accepted.